crates/sui-source-validation/src/lib.rs

```diff
@@ -0,0 +1,252 @@
+                dep.symbol == local_pkg_symbol_str && dep.module_bytes.contains_key(mod_str)
+            }) {
+                missing_modules.push(format!("{}::{}", local_pkg_symbol_str, mod_str))
+            }
+        }
+        missing_modules
+    }
+
+    fn get_module_bytes_map(compiled_package: &CompiledPackage) -> ModuleBytesMap {
+        let mut map: ModuleBytesMap = HashMap::new();
+        compiled_package
+            .deps_compiled_units
+            .iter()
+            .for_each(|(symbol, unit_src)| {
+                let name = unit_src.unit.name();
+                // in the future, this probably needs to specify the compiler version instead of None
+                let bytes = unit_src.unit.serialize(None);
+
+                if let CompiledUnitEnum::Module(m) = unit_src.unit.clone() {
+                    let module_addr: AccountAddress = m.address.into_inner();
+
+                    match map.get_mut(symbol) {
+                        Some((_addr, existing_modules)) => {
+                            existing_modules.insert(name, bytes);
+                        }
+                        None => {
+                            let mut new_map = HashMap::new();
+                            new_map.insert(name, bytes);
+                            map.insert(*symbol, (module_addr, new_map));
+                        }
+                    }
+                }
+            });
+
+        map
+    }
+
+    async fn pkg_for_address(
+        &mut self,
+        addr: &AccountAddress,
+    ) -> Result<SuiRawMovePackage, DependencyVerificationError> {
+        // Move packages are specified with an AccountAddress, but are
+        // fetched from a sui network via sui_getObject, which takes an object ID
+        let obj_id = match ObjectID::from_str(addr.to_string().as_str()) {
+            Ok(id) => id,
+            Err(err) => return Err(DependencyVerificationError::ObjectIdFromAddressFailure(err)),
+        };
+
+        // fetch the Sui object at the address specified for the package in the local resolution table
+        // if future packages with a large set of dependency packages prove too slow to verify,
+        // batched object fetching should be added to the ReadApi & used here
+        let obj_read = match self.rpc_client.get_object(obj_id).await {
+            Ok(raw) => raw,
+            Err(err) => {
+                return Err(DependencyVerificationError::DependencyObjectReadFailure(
+                    err,
+                ))
+            }
+        };
+        let obj = match obj_read.object() {
+            Ok(sui_obj) => sui_obj,
+            Err(err) => return Err(DependencyVerificationError::SuiObjectRefFailure(err)),
+        };
+
+        match obj.data.clone() {
+            SuiRawData::Package(pkg) => Ok(pkg),
+            SuiRawData::MoveObject(move_obj) => Err(
+                DependencyVerificationError::ObjectFoundWhenPackageExpected(obj_id, move_obj),
+            ),
+        }
+    }
```

Comment:
Better to clone after inspecting (You may need to add an ampersand before `obj.data` and/or` some `ref`s to the match pattern variables, I usually rely on the language server to help me there):

```suggestion
        match obj.data {
            SuiRawData::Package(pkg) => Ok(pkg.clone()),
            SuiRawData::MoveObject(move_obj) => Err(
                DependencyVerificationError::ObjectFoundWhenPackageExpected(obj_id, move_obj),
            ),
        }
    }
```

Reply:
what's the difference ?